File: hs/38.hs
-- 932718654
import Euler(getDigitsBase, digitUsage)

nn = 9

buildNum [] = 0
buildNum (x:xs) = x + 10*buildNum xs

genMult0 _ [] _ = error "genMult0: empty"
genMult0 xs (y:ys) n
    | length xs2 == n = buildNum xs2
    | length xs2 > n = 0
    | otherwise = genMult0 xs2 ys n
    where xs2 = getDigitsBase 10 y ++ xs
genMult n x = genMult0 [] [x,x+x..] n

largestPanMult n = maximum $ filter isPandigital $ map (genMult n) [1..z-1]
    where z = 10^(n `div` 2) -- need to generate at least 2 numbers
          isPandigital x = digitUsage (buildNum [1..n]) == digitUsage x

main = putStrLn $ show $ largestPanMult nn

File: py/38.py
# 932718654
import euler
N = 9
m = 0
for a in xrange(1, 10 ** (N/2)):
    n = 0
    s = ''
    while len(s) < N:
        n += 1
        s += str(a * n)
        if len(s) == N and euler.digit_usage(int(s)) == int('1' * 9 + '0', 2):
            if int(s) > m:
                m = int(s)
print m
